# idseq-bench
Benchmark generator for the [IDSeq Portal](https://idseq.net).

So far just a thin wrapper around [InSilicoSeq](https://insilicoseq.readthedocs.io/en/latest/).

# setup
```
pip3 install InSilicoSeq
pip3 install ncbi-acc-download
```

# running
```
python3 main.py
```

This produces zipped fastq files that you can upload via [IDSEQ-CLI](https://github.com/chanzuckerberg/idseq-cli)
to project Benchmarking.


# selecting organisms and chromosomes
Add/modify entries like this in `params.py`
```
    Genome("fungi", "aspergillus_fumigatus",
           [("subspecies", 330879), ("species", 746128), ("genus", 5052), ("family", 1131492)],
           ["NC_007194.1", "NC_007195.1", "NC_007196.1", "NC_007197.1", "NC_007198.1", "NC_007199.1",
            "NC_007200.1", "NC_007201.1"],
           "https://www.ncbi.nlm.nih.gov/genome/18?genome_assembly_id=22576"),
```

# tweaking InSilicoSeq options
Edit [params.py](params.py) or [main.py](main.py) as desired, e.g., to select a different set of error models.

Ensure the output name reflects all chosen params.  If the output data changes,
we expect the output file name to change as well.  It's always a good idea to increment
LOGICAL_VERSION (whole numbers only) after making changes to the code.

TODO:  Random generator seed control.

# interpreting the output
We generate a summary file for each pair of fastqs, indicating read counts per organism.  Each pair
counts as 2 reads, matching InSilicoSeq and IDSeq conventions.

We alter the read id's generated by ISS to satisfy the input requirements of tools like [STAR](https://github.com/alexdobin/STAR).
This requires stripping those `_1` and `_2` pair indicators from all read id's, so that both reads in a pair have the exact same
read ID.  Then, each read ID gets a serial number, and a tag identifying the taxonomic lineage of the organism
that read was sourced from.  This is helpful in tracking reads through complex bioinformatic pipelines and
scoring results.  We assume the pipelines would not cheat by inspecting those tags.
